File: android/core/src/main/kotlin/com/masilotti/bridgecomponents/theme/ThemeComponent.kt
package com.masilotti.bridgecomponents.theme

import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
import com.masilotti.bridgecomponents.shared.Bridgework
import com.masilotti.bridgecomponents.shared.Events
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

class ThemeComponent(
    name: String,
    bridgeDelegate: BridgeDelegate<HotwireDestination>
) : BridgeComponent<HotwireDestination>(name, bridgeDelegate) {

    override fun onReceive(message: Message) {
        val data = message.data<MessageData>() ?: return
        when (data.theme) {
            Theme.LIGHT -> apply(AppCompatDelegate.MODE_NIGHT_NO)
            Theme.DARK -> apply(AppCompatDelegate.MODE_NIGHT_YES)
            null -> apply(MODE_NIGHT_FOLLOW_SYSTEM)
        }
    }

    private fun apply(@AppCompatDelegate.NightMode theme: Int) {
        if (AppCompatDelegate.getDefaultNightMode() != theme) {
            AppCompatDelegate.setDefaultNightMode(theme)
            Bridgework.post(Events.themeDidChange, theme)
        }
    }

    @Serializable
    private data class MessageData(
        val theme: Theme?
    )

    @Serializable
    private enum class Theme {
        @SerialName("light") LIGHT,
        @SerialName("dark") DARK
    }
}

File: ios/components/ThemeComponent.swift
import HotwireNative
import UIKit

public final class ThemeComponent: BridgeComponent {
    override public nonisolated class var name: String { "theme" }

    private var window: UIWindow? {
        let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene
        return scene?.windows.first
    }

    override public func onReceive(message: Message) {
        guard let data: MessageData = message.data() else { return }

        switch data.theme {
        case .light: apply(theme: .light)
        case .dark: apply(theme: .dark)
        default: apply(theme: .unspecified)
        }
    }

    private func apply(theme: UIUserInterfaceStyle) {
        if window?.overrideUserInterfaceStyle != theme {
            window?.overrideUserInterfaceStyle = theme
            Bridgework.post(.themeDidChange, theme)
        }
    }
}

private extension ThemeComponent {
    struct MessageData: Decodable {
        let theme: Theme?
    }
}

private extension ThemeComponent {
    enum Theme: String, Decodable {
        case light
        case dark
    }
}
